TST use global_random_seed in `sklearn/linear_model/tests/test_logistic.py`

DeaMariaLeon opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Towards https://github.com/scikit-learn/scikit-learn/issues/22827

#### What does this implement/fix? Explain your changes.
- I didn't change the tests with a comment about `multi-class` being removed in version 1.8
- These need fixing: - I couldn't make them work with the global random seed `test_LogisticRegression_elastic_net_objective`
`test_LogisticRegressionCV_GridSearchCV_elastic_net`
`test_scores_attribute_layout_elasticnet`
`test_elastic_net_vs_l1_l2`
- For some tests I had to change the parameters to make them work.

#### Any other comments?
Friendly ping to @glemaitre
Also bugging @jeremiedbb  just in case :) 

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
https://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

lorentzenchr commented:
@DeaMariaLeon Please note that `LogisticRegression` makes use of `random_state` only when `solver == ‘sag’, ‘saga’ or ‘liblinear’` (to shuffle the data).

DeaMariaLeon committed:
global-random-seed clean up

DeaMariaLeon committed:
wip

DeaMariaLeon committed:
wip

DeaMariaLeon committed:
wip

DeaMariaLeon committed:
wip

DeaMariaLeon committed:
wip - test_multinomial_log_reg_string_inputs

DeaMariaLeon committed:
wip - tst logistic_cv_sparse

DeaMariaLeon committed:
wip

DeaMariaLeon committed:
test_logistic_regression_solvers

DeaMariaLeon committed:
test_logistic_regression_class_weights

DeaMariaLeon committed:
test_logistic_regression_class_weights

DeaMariaLeon committed:
test_logistic_regression_class_weights - final

DeaMariaLeon committed:
test_logistic_regression_class_weights - again

DeaMariaLeon commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -254,13 +253,16 @@ def test_elasticnet_l1_ratio_err_helpful(LR):
 @pytest.mark.filterwarnings("ignore:.*'multi_class' was deprecated.*:FutureWarning")
 @pytest.mark.parametrize("solver", ["lbfgs", "newton-cg", "sag", "saga"])
-def test_multinomial_binary(solver):
+def test_multinomial_binary(global_random_seed, solver):
```
Maybe this test doesn't need global_random_seed as it will be removed, same for `test_multinomial_binary_probabilites`??

DeaMariaLeon replied:
Remove test now

DeaMariaLeon commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -685,11 +699,13 @@ def test_ovr_multinomial_iris():
+    )
 
-    params = dict(fit_intercept=False, random_state=42)
+    params = dict(C=0.1, fit_intercept=False, random_state=global_random_seed)
```
Is it better to skip setting the `global_random_seed` for the solvers than don't use it, but make the code a bit more complex? or to set it to all of them but make the test take longer?

DeaMariaLeon committed:
test_logistic_regression_solvers_multiclass_unpenalized

DeaMariaLeon committed:
test_logistic_regression_class_weights

DeaMariaLeon committed:
removed test

DeaMariaLeon committed:
Merge remote-tracking branch 'upstream/main' into logistic

DeaMariaLeon committed:
test_predict_iris

DeaMariaLeon committed:
[ci skip] test_concistency_path

DeaMariaLeon committed:
wip

DeaMariaLeon committed:
seed up to test_logistic_regression_multinomial

DeaMariaLeon committed:
seed up to test_logistic_regression_multinomial - clean up

DeaMariaLeon committed:
[ci skip] test_saga_sparse

DeaMariaLeon committed:
test_logreg_l1

DeaMariaLeon committed:
[ci skip] up to test_logistic_regression_cv_refit

DeaMariaLeon committed:
Changed up to test_max_iter

DeaMariaLeon committed:
[ci skip] up to test_elastic_net_coeffs

DeaMariaLeon committed:
[ci skip] test_elastic_net_l1_l2_equivalence

DeaMariaLeon committed:
[ci skip] test_sample_weight_not_modified

DeaMariaLeon committed:
test_liblinear_not_stuck

DeaMariaLeon committed:
Merge remote-tracking branch 'upstream/main' into logistic

DeaMariaLeon commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -147,36 +144,19 @@ def test_predict_3_classes(csr_container):
-    give good accuracy score (>0.95) for the training data.
-    """
+def test_predict_iris(clf, global_random_seed):
+    """Test logistic regression with the iris dataset."""
```
Should the whole test be removed?

DeaMariaLeon committed:
test_lr_cv_scores_differ_when_sample_weight_is_requested

DeaMariaLeon committed:
Removed test

DeaMariaLeon committed:
removed changes to test_LogisticRegression_elastic_net_objective to start again

DeaMariaLeon committed:
Working on test_LogisticRegression_elastic_net_objective

DeaMariaLeon committed:
Adding global_random_seed to test_logistic [all random seeds]
test_predict_2_classes
test_logistic_cv_mock_scorer
test_predict_iris
test_sparsify
test_write_parameters
test_consistency_path
test_liblinear_dual_random_state
test_logistic_cv
test_logistic_cv_multinomial_score
test_multinomial_logistic_regression_string_inputs
test_logistic_cv_sparse
test_logistic_regression_solvers
test_logistic_regression_solvers_multiclass
test_logistic_regression_solvers_multiclass_unpenalized
test_logistic_regression_class_weights
test_logistic_regression_multinomial
test_liblinear_decision_function_zero
test_liblinear_logregcv_sparse
test_saga_sparse
test_logreg_l1
test_logreg_l1_sparse_data
test_logistic_regression_cv_refit
test_logreg_predict_proba_multinomial
test_max_iter
test_warm_start
test_saga_vs_liblinear
test_dtype_match
test_warm_start_converge_LR
test_elastic_net_coeffs
test_elastic_net_l1_l2_equivalence
test_elastic_net_versus_sgd
test_penalty_none
test_scores_attribute_layout_elasticnet
test_multinomial_identifiability_on_iris
test_sample_weight_not_modified
test_liblinear_not_stuck
test_lr_cv_scores_differ_when_sample_weight_is_requested
test_passing_params_without_enabling_metadata_routing

DeaMariaLeon committed:
test_predict_2_classes
test_logistic_cv_mock_scorer
test_predict_iris
test_sparsify
test_write_parameters
test_consistency_path
test_liblinear_dual_random_state
test_logistic_cv
test_logistic_cv_multinomial_score
test_multinomial_logistic_regression_string_inputs
test_logistic_cv_sparse
test_logistic_regression_solvers
test_logistic_regression_solvers_multiclass
test_logistic_regression_solvers_multiclass_unpenalized
test_logistic_regression_class_weights
test_logistic_regression_multinomial
test_liblinear_decision_function_zero
test_liblinear_logregcv_sparse
test_saga_sparse
test_logreg_l1
test_logreg_l1_sparse_data
test_logistic_regression_cv_refit
test_logreg_predict_proba_multinomial
test_max_iter
test_warm_start
test_saga_vs_liblinear
test_dtype_match
test_warm_start_converge_LR
test_elastic_net_coeffs
test_elastic_net_l1_l2_equivalence
test_elastic_net_versus_sgd
test_penalty_none
test_scores_attribute_layout_elasticnet
test_multinomial_identifiability_on_iris
test_sample_weight_not_modified
test_liblinear_not_stuck
test_lr_cv_scores_differ_when_sample_weight_is_requested
test_passing_params_without_enabling_metadata_routing
test_LogisticRegression_elastic_net_objective
test_LogisticRegressionCV_GridSearchCV_elastic_net

DeaMariaLeon committed:
Adding global_random_seed [all random seeds]
test_predict_2_classes
test_logistic_cv_mock_scorer
test_predict_iris
test_sparsify
test_write_parameters
test_consistency_path
test_liblinear_dual_random_state
test_logistic_cv
test_logistic_cv_multinomial_score
test_multinomial_logistic_regression_string_inputs
test_logistic_cv_sparse
test_logistic_regression_solvers
test_logistic_regression_solvers_multiclass
test_logistic_regression_solvers_multiclass_unpenalized
test_logistic_regression_class_weights
test_logistic_regression_multinomial
test_liblinear_decision_function_zero
test_liblinear_logregcv_sparse
test_saga_sparse
test_logreg_l1
test_logreg_l1_sparse_data
test_logistic_regression_cv_refit
test_logreg_predict_proba_multinomial
test_max_iter
test_warm_start
test_saga_vs_liblinear
test_dtype_match
test_warm_start_converge_LR
test_elastic_net_coeffs
test_elastic_net_l1_l2_equivalence
test_elastic_net_versus_sgd
test_penalty_none
test_scores_attribute_layout_elasticnet
test_multinomial_identifiability_on_iris
test_sample_weight_not_modified
test_liblinear_not_stuck
test_lr_cv_scores_differ_when_sample_weight_is_requested
test_passing_params_without_enabling_metadata_routing
test_LogisticRegression_elastic_net_objective
test_LogisticRegressionCV_GridSearchCV_elastic_net

DeaMariaLeon committed:
Reset last two tests

DeaMariaLeon committed:
Adding global_random seed [all random seeds]
test_predict_2_classes
test_logistic_cv_mock_scorer
test_predict_iris
test_sparsify
test_write_parameters
test_consistency_path
test_liblinear_dual_random_state
test_logistic_cv
test_logistic_cv_multinomial_score
test_multinomial_logistic_regression_string_inputs
test_logistic_cv_sparse
test_logistic_regression_solvers
test_logistic_regression_solvers_multiclass
test_logistic_regression_solvers_multiclass_unpenalized
test_logistic_regression_class_weights
test_logistic_regression_multinomial
test_liblinear_decision_function_zero
test_liblinear_logregcv_sparse
test_saga_sparse
test_logreg_l1
test_logreg_l1_sparse_data
test_logistic_regression_cv_refit
test_logreg_predict_proba_multinomial
test_max_iter
test_warm_start
test_saga_vs_liblinear
test_dtype_match
test_warm_start_converge_LR
test_elastic_net_coeffs
test_elastic_net_l1_l2_equivalence
test_elastic_net_versus_sgd
test_penalty_none
test_scores_attribute_layout_elasticnet
test_multinomial_identifiability_on_iris
test_sample_weight_not_modified
test_liblinear_not_stuck
test_lr_cv_scores_differ_when_sample_weight_is_requested
test_passing_params_without_enabling_metadata_routing

DeaMariaLeon committed:
Try to re-start the CI

DeaMariaLeon committed:
Fail on test_LogisticRegression_elastic_net_objective

DeaMariaLeon committed:
Fail on test_LogisticRegression_elastic_net_objective - again

DeaMariaLeon committed:
wip

DeaMariaLeon committed:
restart the ci?

DeaMariaLeon committed:
fix test_LogisticRegressionCV_elasticnet_attribute_shapes

DeaMariaLeon commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -1830,10 +1688,8 @@ def test_LogisticRegressionCV_no_refit(penalty, multi_class):
-# TODO(1.8): remove filterwarnings after the deprecation of multi_class
 # Remove multi_class an change first element of the expected n_iter_.shape from
 # n_classes to 1 (according to the docstring).
-@pytest.mark.filterwarnings("ignore:.*'multi_class' was deprecated.*:FutureWarning")
```
If n_classes to 1 this test fails

DeaMariaLeon committed:
test_predict_iris fix?

DeaMariaLeon committed:
test_dtype_match

DeaMariaLeon committed:
test_predict_iris

DeaMariaLeon committed:
removed unwanted change

DeaMariaLeon committed:
removed liblinear from test_predict_iris

DeaMariaLeon committed:
putting back filter warnings

DeaMariaLeon committed:
adding back removed stuff

DeaMariaLeon committed:
adding back

DeaMariaLeon committed:
added back more stuf

DeaMariaLeon committed:
wip

DeaMariaLeon committed:
Merge remote-tracking branch 'upstream/main' into logistic

DeaMariaLeon committed:
wip

DeaMariaLeon committed:
wip

DeaMariaLeon committed:
test_warm_start_converge_LR

DeaMariaLeon committed:
test_warm_start_converge_LR - better

DeaMariaLeon committed:
wip

DeaMariaLeon committed:
wip

DeaMariaLeon commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -159,19 +157,21 @@ def test_predict_3_classes(csr_container):
+            C=len(iris.data),
+            solver="sag",
+            tol=1e-2,
+            multi_class="ovr",
```
I add `random_state` below

DeaMariaLeon committed:
test_logreg_l1 again & other clean ups

DeaMariaLeon commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -1192,26 +1226,25 @@ def test_logreg_l1():
-    assert_array_almost_equal(lr_saga.coef_, lr_liblinear.coef_)
 
-    # Noise and constant features should be regularized to zero by the l1
-    # penalty
```
Doesn't hold with `global_random_seed`

DeaMariaLeon committed:
test_logistic_regression_cv_refit

DeaMariaLeon committed:
test_elastic_net_versus_sgd

DeaMariaLeon committed:
All changed test [all random seeds]
test_predict_2_classes
test_logistic_cv_mock_scorer
test_predict_iris
test_sparsify
test_write_parameters
test_consistency_path
test_liblinear_dual_random_state
test_logistic_cv
test_logistic_cv_multinomial_score
test_multinomial_logistic_regression_string_inputs
test_logistic_cv_sparse
test_logistic_regression_solvers
test_logistic_regression_solvers_multiclass
test_logistic_regression_solvers_multiclass_unpenalized
test_logistic_regression_class_weights
test_logistic_regression_multinomial
test_liblinear_decision_function_zero
test_liblinear_logregcv_sparse
test_saga_sparse
test_logreg_l1
test_logreg_l1_sparse_data
test_logistic_regression_cv_refit
test_logreg_predict_proba_multinomial
test_max_iter
test_warm_start
test_saga_vs_liblinear
test_dtype_match
test_elastic_net_coeffs
test_elastic_net_l1_l2_equivalence
test_elastic_net_versus_sgd
test_penalty_none
test_scores_attribute_layout_elasticnet
test_multinomial_identifiability_on_iris
test_sample_weight_not_modified
test_liblinear_not_stuck
test_lr_cv_scores_differ_when_sample_weight_is_requested
test_passing_params_without_enabling_metadata_routing
test_LogisticRegression_elastic_net_objective
test_LogisticRegressionCV_GridSearchCV_elastic_net
test_warm_start_converge_LR
test_logreg_l1
test_logistic_regression_cv_refit
test_elastic_net_versus_sgd

DeaMariaLeon committed:
test_logreg_l1_sparse_data

DeaMariaLeon commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -1192,16 +1226,13 @@ def test_logreg_l1():
-    assert_array_almost_equal(lr_saga.coef_[0, -5:], np.zeros(5))
+    assert_allclose(lr_saga.coef_, lr_liblinear.coef_, atol=0.3)
 
 
```
This `test_logreg_l1_sparse_data` doesn't hold with global random seed as is - or couldn't find the right combination. I  wonder if I should do like the previous test.

DeaMariaLeon committed:
test_saga_vs_liblinear

DeaMariaLeon committed:
test_penalty_none

DeaMariaLeon committed:
test_elastic_net_coeffs

DeaMariaLeon committed:
remove random seed from test_scores_attribute_layout_elasticnet

DeaMariaLeon committed:
[ci skip] wip

DeaMariaLeon committed:
test_logreg_l1_sparse_data

DeaMariaLeon committed:
Adding global_random_seed  [all random seeds]
test_predict_2_classes
test_logistic_cv_mock_scorer
test_predict_iris
test_sparsify
test_write_parameters
test_consistency_path
test_liblinear_dual_random_state
test_logistic_cv
test_logistic_cv_multinomial_score
test_multinomial_logistic_regression_string_inputs
test_logistic_cv_sparse
test_logistic_regression_solvers
test_logistic_regression_solvers_multiclass
test_logistic_regression_solvers_multiclass_unpenalized
test_logistic_regression_class_weights
test_logistic_regression_multinomial
test_liblinear_decision_function_zero
test_liblinear_logregcv_sparse
test_saga_sparse
test_logreg_l1
test_logistic_regression_cv_refit
test_logreg_predict_proba_multinomial
test_max_iter
test_warm_start
test_dtype_match
test_elastic_net_l1_l2_equivalence
test_elastic_net_versus_sgd
test_multinomial_identifiability_on_iris
test_sample_weight_not_modified
test_liblinear_not_stuck
test_lr_cv_scores_differ_when_sample_weight_is_requested
test_passing_params_without_enabling_metadata_routing
test_LogisticRegression_elastic_net_objective
test_LogisticRegressionCV_GridSearchCV_elastic_net
test_warm_start_converge_LR
test_elastic_net_versus_sgd
test_saga_vs_liblinear
test_penalty_none
test_elastic_net_coeffs
test_logreg_l1_sparse_data

DeaMariaLeon committed:
removed global random seed from test_logistic_regression_solvers_multiclass

DeaMariaLeon committed:
forgot this

DeaMariaLeon committed:
All seeds [all random seeds]
test_predict_2_classes
test_logistic_cv_mock_scorer
test_predict_iris
test_sparsify
test_write_parameters
test_consistency_path
test_liblinear_dual_random_state
test_logistic_cv
test_logistic_cv_multinomial_score
test_multinomial_logistic_regression_string_inputs
test_logistic_cv_sparse
test_logistic_regression_solvers
test_logistic_regression_solvers_multiclass_unpenalized
test_logistic_regression_class_weights
test_logistic_regression_multinomial
test_liblinear_decision_function_zero
test_liblinear_logregcv_sparse
test_saga_sparse
test_logreg_l1
test_logistic_regression_cv_refit
test_logreg_predict_proba_multinomial
test_max_iter
test_warm_start
test_dtype_match
test_elastic_net_l1_l2_equivalence
test_elastic_net_versus_sgd
test_multinomial_identifiability_on_iris
test_sample_weight_not_modified
test_liblinear_not_stuck
test_lr_cv_scores_differ_when_sample_weight_is_requested
test_passing_params_without_enabling_metadata_routing
test_LogisticRegression_elastic_net_objective
test_LogisticRegressionCV_GridSearchCV_elastic_net
test_warm_start_converge_LR
test_elastic_net_versus_sgd
test_saga_vs_liblinear
test_penalty_none
test_elastic_net_coeffs
test_logreg_l1_sparse_data

glemaitre commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -560,24 +569,24 @@ def test_logistic_cv_multinomial_score(scoring, multiclass_agg_list):
 
 
-def test_multinomial_logistic_regression_string_inputs():
+def test_multinomial_logistic_regression_string_inputs(global_random_seed):
```
Basically you can add a small docstring explaining what the test is doing.

glemaitre commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -82,19 +82,17 @@ def check_predictions(clf, X, y):
 
 
-def test_logistic_cv_mock_scorer():
+def test_logistic_cv_mock_scorer(global_random_seed):
```
You can add a small docstring explaining what we are testing here.

DeaMariaLeon replied:
Thanks for the feedback

glemaitre commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -1661,6 +1715,7 @@ def test_elastic_net_vs_l1_l2(C):
     assert gs.score(X_test, y_test) >= l2_clf.score(X_test, y_test)
 
 
+# Fix this one:
```
You can write a `# FIXME: ...` and mention that we fixed the random seed to make the test pass.

glemaitre reviewed (commented):
OK it looks good. A couple of comments.

DeaMariaLeon commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -159,19 +157,21 @@ def test_predict_3_classes(csr_container):
         LogisticRegression(C=len(iris.data), solver="lbfgs"),
```
I didn't see any convergence warning after removing the warning filter. I still added the `max_inter=200` (it's `max_iter` right? unless I'm missing a parameter name).

DeaMariaLeon committed:
test_predict_iris

DeaMariaLeon committed:
test_logistic_regression_solvers

DeaMariaLeon committed:
After feedback [all random seeds]
test_predict_2_classes
test_logistic_cv_mock_scorer
test_predict_iris
test_sparsify
test_write_parameters
test_consistency_path
test_liblinear_dual_random_state
test_logistic_cv
test_logistic_cv_multinomial_score
test_multinomial_logistic_regression_string_inputs
test_logistic_cv_sparse
test_logistic_regression_solvers
test_logistic_regression_solvers_multiclass_unpenalized
test_logistic_regression_class_weights
test_logistic_regression_multinomial
test_liblinear_decision_function_zero
test_liblinear_logregcv_sparse
test_saga_sparse
test_logreg_l1
test_logistic_regression_cv_refit
test_logreg_predict_proba_multinomial
test_max_iter
test_warm_start
test_dtype_match
test_elastic_net_l1_l2_equivalence
test_elastic_net_versus_sgd
test_multinomial_identifiability_on_iris
test_sample_weight_not_modified
test_liblinear_not_stuck
test_lr_cv_scores_differ_when_sample_weight_is_requested
test_passing_params_without_enabling_metadata_routing
test_LogisticRegression_elastic_net_objective
test_LogisticRegressionCV_GridSearchCV_elastic_net
test_warm_start_converge_LR
test_elastic_net_versus_sgd
test_saga_vs_liblinear
test_penalty_none
test_elastic_net_coeffs
test_logreg_l1_sparse_data

DeaMariaLeon commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -1633,25 +1679,31 @@ def test_elastic_net_l1_l2_equivalence(C, penalty, l1_ratio):
 
 @pytest.mark.parametrize("C", [0.001, 1, 100, 1e6])
-def test_elastic_net_vs_l1_l2(C):
+def test_elastic_net_vs_l1_l2(global_random_seed, C):
```
I ended up removing its `global_random_seed`.

DeaMariaLeon commented:
I _think_ I addressed all the comments.

glemaitre reviewed (approved):
LGTM.

glemaitre commented:
Thanks @DeaMariaLeon We can revisit the tests for which it is not obvious how we can fix it.
